File: csharp_WordCount/WordCountTests/WordCountTests.cs
using System;
using System.IO;
using System.Security;
using System.Collections;
using System.Collections.Generic;
using Microsoft.VisualStudio.TestTools.UnitTesting;

using ConsoleApp;

namespace WordCountTests;

// Mock FileManager able to create StreamReaders compatible with WordCountsFromFile
public class StringFileManager : IFileManager {
    private string fileContent;

    public StringFileManager(string fileContent) {
        this.fileContent = fileContent;
    }
    
    public StreamReader OpenText(string path) {
        // Simply keep the stream in memory, this way no actual file I/O is required to run any tests
        var stream = new MemoryStream();
        var writer = new StreamWriter(stream);
        writer.Write(fileContent);
        writer.Flush();
        stream.Position = 0;
        return new StreamReader(stream);
    }
}

// Mock FileManager for throwing exceptions to test for proper file handling
public class ExceptionFileManager : IFileManager {
    private Exception exception;
    
    public ExceptionFileManager(Exception exception) {
        this.exception = exception;
    }
    
    public StreamReader OpenText(string path) {
        throw exception;
    }
}

[TestClass]
public class WordCountTests {
    private void TestFromString(string fileContent, IDictionary<string, int> expectedCounts) {
        IFileManager stringFileManager = new StringFileManager(fileContent);
        var counter = new WordCounter();
        var wordCounts = counter.WordCountsFromFile(String.Empty, stringFileManager);
        CollectionAssert.AreEqual((ICollection)expectedCounts, (ICollection)wordCounts);
    }

    [TestMethod]
    public void TestSingleWord() {
        IDictionary<string, int> expected = new Dictionary<string, int>() {{"something", 1}};
        TestFromString("something", expected);
    }

    [TestMethod]
    public void TestSingleString() {
        IDictionary<string, int> expected = new Dictionary<string, int>() {
            {"Go", 1},
            {"do", 2},
            {"that", 2},
            {"thing", 1},
            {"you", 1},
            {"so", 1},
            {"well", 1},
        };
        TestFromString("Go do that thing that you do so well\n", expected);
    }

    [TestMethod]
    public void TestMultipleStrings() {
        IDictionary<string, int> expected = new Dictionary<string, int>() {
            {"Go", 3},
            {"do", 6},
            {"that", 6},
            {"thing", 3},
            {"you", 3},
            {"so", 3},
            {"well", 3},
        };
        TestFromString("Go do that thing that you do so well\nGo do that thing that you do so well\nGo do that thing that you do so well\n", expected);
    }

    [TestMethod]
    public void TestEmptyStrings() {
        IDictionary<string, int> expected = new Dictionary<string, int>();
        TestFromString("", expected);
        TestFromString("\n", expected);
        TestFromString("\r\n", expected);
        TestFromString("          ", expected);
        TestFromString("   \n\r\n", expected);
    }
}

[TestClass]
public class ProgramManagerTests {
    // Ensure that the program performs exception handling correctly - i.e. correct status code is returned

    [TestMethod]
    public void TestExceptionHandling() {
        List<Exception> exceptions = new List<Exception>() {
            new UnauthorizedAccessException(),
            new ArgumentException(),
            new ArgumentNullException(),
            new PathTooLongException(),
            new FileNotFoundException(),
            new NotSupportedException(),
        };

        foreach (var exception in exceptions) {
            IFileManager exceptionFileManager = new ExceptionFileManager(exception);
            var programManager = new ProgramManager(exceptionFileManager);
            Tuple<int, string> result = programManager.Run("some path");
            Assert.AreEqual(result.Item1, 1);
        }
    }

    [TestMethod]
    public void TestSuccessfulWordCount() {
        IFileManager stringFileManager = new StringFileManager("some string");
        var programManager = new ProgramManager(stringFileManager);
        Tuple<int, string> result = programManager.Run("some path");
        Assert.AreEqual(result.Item1, 0);
    }
}

File: fsharp_WordCount/WordCountTests/WordCountTests.fs
namespace fsharp_WordCount.Tests

open System
open System.IO
open System.Security
open Microsoft.VisualStudio.TestTools.UnitTesting

open WordCount


[<TestClass>]
type WordCountsTests () =

    let seqEquals expected result = Seq.compareWith Operators.compare expected result = 0

    [<TestMethod>]
    member this.TestSingleWord () = 
        let expected = seq {("something", 1)}
        let result = wordCounts (seq {"something\n"})
        Assert.IsTrue(seqEquals expected result)

    [<TestMethod>]
    member this.TestSingleString () = 
        let expected = seq {
            ("Go", 1);
            ("do", 2);
            ("that", 2);
            ("thing", 1);
            ("you", 1);
            ("so", 1);
            ("well", 1);
        }
        let result = wordCounts (seq {"Go do that thing that you do so well\n"})
        Assert.IsTrue(seqEquals expected result)

    [<TestMethod>]
    member this.TestMultipleStrings () = 
        let expected = seq {
            ("Go", 3);
            ("do", 6);
            ("that", 6);
            ("thing", 3);
            ("you", 3);
            ("so", 3);
            ("well", 3);
        }
        let result = wordCounts (seq {
            "Go do that thing that you do so well\n";
            "Go do that thing that you do so well\n";
            "Go do that thing that you do so well\n";
        })
        Assert.IsTrue(seqEquals expected result)

    [<TestMethod>]
    member this.TestEmptyStrings () = 
        let expected = Seq.empty
        let result = wordCounts (seq {""; "\n"; "\r\n"; "     "; "  \n\r\n"})
        Assert.IsTrue(seqEquals expected result)


[<TestClass>]
type FileWordCountsTests () =

    let fileWordCountsAlwaysRaise ex =
        // mocking function replacing File.ReadLines - to instead always throw a given exception
        let exFunc = (fun _path -> raise ex)
        let (status_code, _) = fileWordCounts exFunc "some path"
        Assert.AreEqual(status_code, 1)

    [<TestMethod>]
    member this.TestExceptionHandling () =
        // Ensure that an exception possibly thrown by File.ReadLines are properly handled
        let possibleExceptions: seq<Exception> = seq {
            new ArgumentException();
            new ArgumentNullException();
            new FileNotFoundException();
            new IOException();
            new PathTooLongException();
            new SecurityException();
            new UnauthorizedAccessException();
        }
        Seq.iter fileWordCountsAlwaysRaise possibleExceptions

    [<TestMethod>]
    member this.TestSuccessfulWordCount () =
        // Ensure that status code is correctly set when file is read correctly
        let successFunc = (fun (path: string) -> seq {path})
        let (status_code, _) = fileWordCounts successFunc "some path"
        Assert.AreEqual(status_code, 0)
